### DISCUSSION TITLE
Power save design

### DISCUSSION THREAD

**Discussion by TomK57** [jdohm/ZP_Muelleimer#4] [Ideas]
Ressourcen effizient einsetzen gehört zu jedem technischen Projekt. Deshalb die Überlegung zum Stromverbrauch des Projekts.

Der ESP8266 allein braucht im Normalbetrieb, ohne Senden, laut Datenblatt durchschnittlich 80mA. Dazu kommt der Stromverbrauch des USB Chip, die Verlustleistung des 3,3V Reglers und des USB Netzteils und die RGB Elektronik im Leerlauf. In Summe ca. 100mA. Das sind bei 5V Eingangsspannung ca. 0,5W, eigentlich zu vernachlässigen, oder?
Wenn man aber regelmässig den Timeserver abfragt und das ESP8266 Modem dauerhaft betreibt, steigt der Stromverbrauch entsprechend. Im Sendebetrieb fliessen kurzzeitig bis zu 400mA. Das sind immer noch geringe Leistungen im Vergleich zum Rest eines normalen Haushalts, aber wie der Volksmund treffend weiss: Kleinvieh macht auch Mist.

Grundsätzlich sollte man immer sparsam mit Ressourcen umgehen und wo sich etwas einfach einsparen lässt, sollte man es auch tun. Wie gesagt, auch kleine Beträge summieren sich.

In unserem Fall untersuchen wir deshalb den Use Case nach seinen wesentlichen Anforderungen, um daraus Einsparpotential abzuleiten.
1. Die Einträge im Müllkalenders sind tageweise
2. Der Kalendertag ändert sich alle 24h

Folgerung
1. Prüfung des aktuellen Tages braucht nur einmal alle 24h zu erfolgen.
2. Das System kann bis auf die Zeit für eine Zeitabfrage schlafen, die Anzeige bleibt solange konstant.  

Überlegung zum Schlafen:
Der ESP8266 kennt verschiedene Sleep Modi, darunter einen Deep Sleep Modus, bei dem Modem und CPU abgeschaltet sind und nur eine interne RealTimeClock weiterläuft, die über einen GPIO Pin das System wieder wecken kann. Die Schlafdauer ist einstellbar.
Im Deep Sleep Modus verbraucht der D1 Mini nur noch 17mA. Das ist doch signifikant. Absolut ist der Wert nicht beeindruckend aber prozentual können wir ca. 80% des Stromverbrauchs einsparen.

Was ist zu beachten?
1. Im Nutzfall, "Müll soll raus" signalisieren, müssen die LED trotz Schlafmodus weiterleuchten
2. Am Morgen des Tages für die Benachrichtigung, sollen die LEDs leuchten.

zu 1. Der LED Treiber behält die letzte Einstellung bei, solange er mit Spannung versorgt ist und keine neuen Daten kommen. Die Stromaufnahme liegt bei 5mA, wenn alle LED aus sind.
zu 2. Damit das Wecken am frühen Morgen stattfindet, müsste die Schlafdauer in Abhängigkeit vom ersten Einschlafen nach Inbetriebnahme erfolgen und nach dem ersten Aufwachen auf 24h angepasst werden. Das würde den Einbau eines zusätzlichen "Gedächtnisses" erfordern, weil der ESP8266 im Deep Sleep Modus sein RAM-Gedächtnis verliert.
Prinzipiell ginge das, weil der EPS8266 Flash Speicher hat, der für Daten verwendet werden kann. Auf der anderen Seite würde das einiges an Software Komplexität erfordern, die nicht gewollt ist.
Was ist das worst case Szenario für das Schlafkonzept? Das System geht kurz vor Mitternacht schlafen und der nächste Tag ist ein "Müll rausbringen" Tag. Wenn wir die Schlafdauer von Beginn an konstant auf 6h einstellen, wird der Nutzer spätestens um 6:00 informiert. Das sollte für unsere Anwender ausreichend sein. Eine Schlafdauer von 6h bedeutet viermal in 24h Aufwachen für maximal 10-20s, das ist wirklich vernachlässigbar.     

Wie sieht das SW Design in der Arduino IDE schliesslich aus?

Funktion setup()
- establish connection to WLAN
- set time and date for system time 
- check calendar file
- signal result
- put system to deep sleep for 6h

Funktion loop()
- do nothing

Einfacher geht's nicht!

**Comment by jdohm** [comment]
Hört sich gut an, ich würde sagen: Ausprobieren! Wenn wir ein Beispiel haben können wir auch gut schauen ob die Komplexität für Einsteiger gut machbar ist. Hört sich aber fürs erste machbar an. 

Ein "Problem" sehe ich darin, das so das System für Tage tot wirkt. Deswegen hatte ich bei meinem immer ein Blinken implementiert, damit ersichtlich ist, das Gerät läuft aktuell und ist nicht etwa eingefroren. Auf diese Zusatzfunktion kann man aber eventuell verzichten, zumal eine Blinkende LED (ganz von der Ansteuerung abgesehen) auch schon einiges an Leistung aufnimmt.
